SDWebImage/SDAnimatedImage.m

```diff
@@ -0,0 +1,158 @@
+        if (result.range.location >= 3) {
+            scale = [string substringWithRange:NSMakeRange(result.range.location + 1, result.range.length - 2)].doubleValue;
+        }
+    }];
+    
+    return scale;
+}
+
+@interface SDAnimatedImage ()
+
+@property (nonatomic, strong) id<SDWebImageAnimatedCoder> coder;
+@property (nonatomic, assign, readwrite) NSUInteger animatedImageLoopCount;
+@property (nonatomic, assign, readwrite) NSUInteger animatedImageFrameCount;
+@property (nonatomic, assign, readwrite) SDImageFormat animatedImageFormat;
+@property (nonatomic, assign) BOOL animatedImageLoopCountCheck;
+@property (nonatomic, assign) BOOL animatedImageFrameCountChecked;
+
+#if SD_MAC
+@property (nonatomic, assign) CGFloat scale;
+#endif
+
+@end
+
+@implementation SDAnimatedImage
+
+#pragma mark - UIImage override method
++ (instancetype)imageWithContentsOfFile:(NSString *)path {
+    return [[self alloc] initWithContentsOfFile:path];
+}
+
++ (instancetype)imageWithData:(NSData *)data {
+    return [[self alloc] initWithData:data];
+}
+
++ (instancetype)imageWithData:(NSData *)data scale:(CGFloat)scale {
+    return [[self alloc] initWithData:data scale:scale];
+}
+
+- (instancetype)initWithContentsOfFile:(NSString *)path {
+    NSData *data = [NSData dataWithContentsOfFile:path];
+    return [self initWithData:data scale:SDImageScaleFromPath(path)];
+}
+
+- (instancetype)initWithData:(NSData *)data {
+    return [self initWithData:data scale:1];
+}
+
+- (instancetype)initWithData:(NSData *)data scale:(CGFloat)scale {
+    @autoreleasepool {
+        for (id<SDWebImageCoder>coder in [SDWebImageCodersManager sharedInstance].coders) {
+            if ([coder conformsToProtocol:@protocol(SDWebImageAnimatedCoder)]) {
+                if ([coder canDecodeFromData:data]) {
+                    id<SDWebImageAnimatedCoder> animatedCoder = [[[coder class] alloc] initWithAnimatedImageData:data];
+                    if (!animatedCoder) {
+                        // check next coder
+                        continue;
+                    } else {
+                        self.coder = animatedCoder;
+                        break;
+                    }
+                }
+            }
+        }
+        if (!self.coder) {
+            return nil;
+        }
+        UIImage *image = [self.coder animatedImageFrameAtIndex:0];
+        if (!image) {
+            return nil;
+        }
+#if SD_MAC
+        self = [super initWithCGImage:image.CGImage size:NSZeroSize];
+#else
+        self = [super initWithCGImage:image.CGImage scale:scale orientation:image.imageOrientation];
+#endif
+        if (!self) {
+            return nil;
+        }
+        SDImageFormat format = [NSData sd_imageFormatForImageData:data];
+        self.animatedImageFormat = format;
+    }
+    return self;
+}
+
+#pragma mark - NSSecureCoding
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    NSNumber *scale = [aDecoder decodeObjectOfClass:[NSNumber class] forKey:@"scale"];
```

Comment:
To avoid using hard-code string, keys can be generated from selector like `NSStringFromSelector(@selector(scale))`

Reply:
Good. I loss this during a git reset :)